examples/rn-multiple-realms/frontend/app/providers/MovieProvider.tsx

```diff
@@ -0,0 +1,175 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import React, {
+  ReactNode,
+  createContext,
+  useCallback,
+  useContext,
+  useMemo,
+  useState,
+} from 'react';
+import type Realm from 'realm';
+import {useQuery, useRealm} from '@realm/react';
+
+import {Movie} from '../models/Movie';
+import {PrivateContent} from '../models/PrivateContent';
+
+/**
+ * Values available to consumers of the `MovieContext`.
+ */
+type MovieContextType = {
+  /**
+   * All movies grouped by category.
+   */
+  movieSections: {
+    category: string;
+    // `My List` is a `Realm.List<Movie>` while the other
+    // movie categories will yield `Realm.Results<Movie>`.
+    movies: Realm.Results<Movie> | Realm.List<Movie>;
+  }[];
+  /**
+   * The movie selected to view more information about.
+   */
+  selectedMovie: Movie | null;
+  setSelectedMovie: (movie: Movie) => void;
+  addToMyList: (movie: Movie) => void;
+  removeFromMyList: (movie: Movie) => void;
+  existsInMyList: (movie: Movie) => boolean;
+};
+
+/**
+ * The movie context with initial values.
+ */
+const MovieContext = createContext<MovieContextType>({
+  movieSections: [],
+  selectedMovie: null,
+  setSelectedMovie: () => {},
+  addToMyList: () => {},
+  removeFromMyList: () => {},
+  existsInMyList: () => false,
+});
+
+type MovieProviderProps = {
+  children: ReactNode;
+};
+
+/**
+ * Queries and provides the relevant movies using `@realm/react`, as well as
+ * providing functions for updating `My List`.
+ */
+export function MovieProvider({children}: MovieProviderProps) {
+  const realm = useRealm();
+
+  // The selected movie will decide which movie is shown on the
+  // `MovieInfoScreen` when a movie item is pressed from the list.
+  const [selectedMovie, setSelectedMovie] = useState<Movie | null>(null);
+
+  // Query movies of certain genres. (A helper `query()` function is
+  // defined to simplify the argument passed to `useQuery()`).
+  const query = (movies: Realm.Results<Movie>, genre: string) =>
+    movies.filtered(`'${genre}' IN genres`);
+  const action = useQuery(Movie, movies => query(movies, 'Action'));
```

Comment:
I'm wondering with this setup if when one of these `useQuery` hooks triggers a re-render, if the entire application will re-render, even if the data isn't currently in use.

Reply:
Yeah in this case, the children of this provider will rerender. Once the movies have been synced, the only data being updated however will be the `My List`. I tried a bunch of variants in this file to try to notice speed differences. Using e.g. multiple `useQuery()`s with individual filters (one for each genre) were better than using one `useQuery()` and calling `.filtered()` for each genre.

But more to your point, since the `MoviesNavigator` is the only component using the context value, one thing I can try to do is to move the `MovieProvider` to only wrap that, rather than wrapping `RootNavigator` 👍